src/main/java/duke/Storage.java

```diff
@@ -0,0 +1,97 @@
+package duke;
+
+import java.io.FileNotFoundException;
+import java.io.FileWriter;
+import java.io.File;
+import java.io.IOException;
+import java.util.Scanner;
+
+public class Storage {
+    private static final String FILE_PATH = "data.txt";
+
+    public static void loadData(TaskManager taskManager) throws FileNotFoundException {
+        String filePath = FILE_PATH;
+        File data = new File(filePath);
+        Scanner scanner = new Scanner(data);
+
+        while(scanner.hasNextLine()) {
+            String line = scanner.nextLine();
+            try {
+                parseLine(line, taskManager);
+            } catch (DukeException.IllegalReadFromSaveData e) {
+                System.out.println("Oops, something went wrong, unable to read from last save");
+                return;
+            } catch (DukeException.IllegalMarkTargetException e) {
+                System.out.println("Oops, something went wrong, unable to read from last save");
+                return;
+            }
+        }
+        System.out.println("Successfully loaded your last save of tasks to do!");
+        taskManager.listTasks();
+    }
+
+    private static void parseLine(String line, TaskManager taskManager)
+            throws DukeException.IllegalReadFromSaveData, DukeException.IllegalMarkTargetException {
+        String[] arrOfLine = line.split("\\|");
+        int length = arrOfLine.length;
+
+        if (length < 3 || length > 4) {
+            throw new DukeException.IllegalReadFromSaveData();
+        }
+        String taskType = arrOfLine[0];
+
+        switch (taskType) {
+        case "T":
+            if (length == 4) {
+                throw new DukeException.IllegalReadFromSaveData();
+            }
+            createNewTodo(arrOfLine, taskManager);
+            break;
+        case "D":
+            if (length == 3) {
+                throw new DukeException.IllegalReadFromSaveData();
+            }
+            createNewDeadline(arrOfLine, taskManager);
+            break;
+        case "E":
+            if (length == 3) {
+                throw new DukeException.IllegalReadFromSaveData();
+            }
+            createNewEvent(arrOfLine, taskManager);
+            break;
```

Comment:
In such cases, you might want to consider setting the length parameters as variables instead, as in this context, it may be good in attempting to eliminate magic numbers.